File: crates/encompute-openfhe/cpp/binfhe.cc
#include "binfhe.h"

#include <sstream>
#include <stdexcept>

#include "binfhecontext.h"
#include "binfhecontext-ser.h"
#include "common.h"
#include "encompute-openfhe/src/binfhe.rs.h"

namespace encompute_openfhe {

using lbcrypto::BinFHEContext;
using lbcrypto::LWECiphertext;

struct BinContextImpl {
  BinFHEContext cc;
  bool keys = false;
};

struct BinCiphertextImpl {
  LWECiphertext ct;
};

BinContext::BinContext(std::unique_ptr<BinContextImpl> i) : impl(std::move(i)) {}
BinContext::~BinContext() {
  std::lock_guard<std::mutex> g(openfhe_mutex());
  impl.reset();
}
BinCiphertext::BinCiphertext(std::unique_ptr<BinCiphertextImpl> i) : impl(std::move(i)) {}
BinCiphertext::~BinCiphertext() {
  std::lock_guard<std::mutex> g(openfhe_mutex());
  impl.reset();
}

static std::unique_ptr<BinCiphertext> wrap(LWECiphertext ct) {
  auto i = std::make_unique<BinCiphertextImpl>();
  i->ct = std::move(ct);
  return std::make_unique<BinCiphertext>(std::move(i));
}

static lbcrypto::BINFHE_METHOD method_of(const std::string& name) {
  return name.find("LMKCDEY") != std::string::npos ? lbcrypto::LMKCDEY : lbcrypto::GINX;
}

lbcrypto::BINFHE_PARAMSET bin_paramset(const std::string& name) {
  if (name == "STD128") return lbcrypto::STD128;
  if (name == "STD128Q") return lbcrypto::STD128Q;
  if (name == "STD128_LMKCDEY") return lbcrypto::STD128_LMKCDEY;
  throw std::runtime_error("unsupported BinFHE parameter set " + name +
                           " (vetted: STD128, STD128Q)");
}

std::unique_ptr<BinContext> bin_new_context(rust::Str paramset) {
  std::lock_guard<std::mutex> g(openfhe_mutex());
  auto i = std::make_unique<BinContextImpl>();
  i->cc.GenerateBinFHEContext(bin_paramset(std::string(paramset)), method_of(std::string(paramset)));
  return std::make_unique<BinContext>(std::move(i));
}

uint64_t bin_lwe_n(const BinContext& ctx) {
  std::lock_guard<std::mutex> g(openfhe_mutex());
  return const_cast<BinFHEContext&>(ctx.impl->cc).GetParams()->GetLWEParams()->Getn();
}

uint64_t bin_lwe_q(const BinContext& ctx) {
  std::lock_guard<std::mutex> g(openfhe_mutex());
  return const_cast<BinFHEContext&>(ctx.impl->cc).GetParams()->GetLWEParams()->Getq().ConvertToInt();
}

template <typename T>
static T deserialize(rust::Slice<const uint8_t> bytes, const char* what) {
  T obj;
  try {
    std::string s(reinterpret_cast<const char*>(bytes.data()), bytes.size());
    std::istringstream in(s);
    lbcrypto::Serial::Deserialize(obj, in, lbcrypto::SerType::BINARY);
  } catch (const std::exception& e) {
    throw std::runtime_error(std::string("malformed ") + what + ": " + e.what());
  }
  if (!obj) throw std::runtime_error(std::string("malformed ") + what);
  return obj;
}

void bin_load_keys(BinContext& ctx, rust::Slice<const uint8_t> refresh,
                   rust::Slice<const uint8_t> switching) {
  std::lock_guard<std::mutex> g(openfhe_mutex());
  auto bs = deserialize<lbcrypto::RingGSWACCKey>(refresh, "BinFHE refresh key");
  auto ks = deserialize<lbcrypto::LWESwitchingKey>(switching, "BinFHE switching key");
  lbcrypto::RingGSWBTKey key;
  key.BSkey = bs;
  key.KSkey = ks;
  ctx.impl->cc.BTKeyLoad(key);
  ctx.impl->keys = true;
}

std::unique_ptr<BinCiphertext> bin_load(const BinContext& ctx, rust::Slice<const uint8_t> bytes) {
  std::lock_guard<std::mutex> g(openfhe_mutex());
  auto ct = deserialize<LWECiphertext>(bytes, "BinFHE ciphertext");
  auto p = const_cast<BinFHEContext&>(ctx.impl->cc).GetParams()->GetLWEParams();
  if (ct->GetLength() != p->Getn() || ct->GetModulus() != p->Getq()) {
    throw std::runtime_error("ciphertext is for another parameter set");
  }
  return wrap(std::move(ct));
}

rust::Vec<uint8_t> bin_store(const BinCiphertext& ct) {
  std::lock_guard<std::mutex> g(openfhe_mutex());
  std::ostringstream out;
  lbcrypto::Serial::Serialize(ct.impl->ct, out, lbcrypto::SerType::BINARY);
  auto s = out.str();
  rust::Vec<uint8_t> v;
  v.reserve(s.size());
  for (char c : s) v.push_back(static_cast<uint8_t>(c));
  return v;
}

std::unique_ptr<BinCiphertext> bin_gate(const BinContext& ctx, uint8_t gate,
                                        const BinCiphertext& a, const BinCiphertext& b) {
  static const lbcrypto::BINGATE gates[] = {lbcrypto::OR,  lbcrypto::AND, lbcrypto::NOR,
                                            lbcrypto::NAND, lbcrypto::XOR, lbcrypto::XNOR};
  if (gate > 5) throw std::runtime_error("unknown BinFHE gate");
  std::lock_guard<std::mutex> g(openfhe_mutex());
  if (!ctx.impl->keys) throw std::runtime_error("no bootstrapping keys loaded");
  return wrap(ctx.impl->cc.EvalBinGate(gates[gate], a.impl->ct, b.impl->ct));
}

std::unique_ptr<BinCiphertext> bin_not(const BinContext& ctx, const BinCiphertext& a) {
  std::lock_guard<std::mutex> g(openfhe_mutex());
  return wrap(ctx.impl->cc.EvalNOT(a.impl->ct));
}

std::unique_ptr<BinCiphertext> bin_constant(const BinContext& ctx, bool value) {
  std::lock_guard<std::mutex> g(openfhe_mutex());
  return wrap(ctx.impl->cc.EvalConstant(value));
}

std::unique_ptr<BinCiphertext> bin_clone(const BinCiphertext& ct) {
  std::lock_guard<std::mutex> g(openfhe_mutex());
  return wrap(std::make_shared<lbcrypto::LWECiphertextImpl>(*ct.impl->ct));
}

}  // namespace encompute_openfhe

File: crates/encompute-openfhe/src/binfhe.rs
//! OpenFHE BinFHE, evaluator side: Boolean gates on LWE ciphertexts with
//! the client's bootstrapping keys. Encompute's integer semantics are built
//! on these gates in `encompute-openfhe-exact`; nothing here knows about
//! integers. No key generation, encryption or decryption.

use encompute_ir::{Code, Error, Result};

#[allow(unsafe_code)]
#[cxx::bridge(namespace = "encompute_openfhe")]
mod ffi {
    unsafe extern "C++" {
        include!("binfhe.h");

        type BinContext;
        type BinCiphertext;

        fn bin_new_context(paramset: &str) -> Result<UniquePtr<BinContext>>;
        fn bin_lwe_n(ctx: &BinContext) -> Result<u64>;
        fn bin_lwe_q(ctx: &BinContext) -> Result<u64>;
        fn bin_load_keys(ctx: Pin<&mut BinContext>, refresh: &[u8], switching: &[u8])
            -> Result<()>;
        fn bin_load(ctx: &BinContext, bytes: &[u8]) -> Result<UniquePtr<BinCiphertext>>;
        fn bin_store(ct: &BinCiphertext) -> Result<Vec<u8>>;
        fn bin_gate(
            ctx: &BinContext,
            gate: u8,
            a: &BinCiphertext,
            b: &BinCiphertext,
        ) -> Result<UniquePtr<BinCiphertext>>;
        fn bin_not(ctx: &BinContext, a: &BinCiphertext) -> Result<UniquePtr<BinCiphertext>>;
        fn bin_constant(ctx: &BinContext, value: bool) -> Result<UniquePtr<BinCiphertext>>;
        fn bin_clone(ct: &BinCiphertext) -> Result<UniquePtr<BinCiphertext>>;
    }
}

fn err(e: cxx::Exception) -> Error {
    let m = e.what();
    let code = if m.contains("another parameter set") {
        Code::WrongParameters
    } else if m.contains("malformed") {
        Code::Envelope
    } else {
        Code::Backend
    };
    Error::new(code, format!("OpenFHE BinFHE: {m}"))
}

/// A two-input bootstrapped gate.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Gate {
    Or = 0,
    And = 1,
    Nor = 2,
    Nand = 3,
    Xor = 4,
    Xnor = 5,
}

/// A BinFHE context with (once loaded) the client's bootstrapping keys.
pub struct BinContext {
    inner: cxx::UniquePtr<ffi::BinContext>,
}

/// One encrypted bit.
pub struct BinCiphertext {
    inner: cxx::UniquePtr<ffi::BinCiphertext>,
}

// SAFETY: every shim function, including the wrappers' destructors, holds
// the global OpenFHE mutex (cpp/common.h), so the objects may move between
// threads and be shared by reference.
#[allow(unsafe_code)]
unsafe impl Send for BinContext {}
#[allow(unsafe_code)]
unsafe impl Sync for BinContext {}
#[allow(unsafe_code)]
unsafe impl Send for BinCiphertext {}
#[allow(unsafe_code)]
unsafe impl Sync for BinCiphertext {}

impl BinContext {
    /// A context for a vetted parameter set (`STD128` or `STD128Q`, GINX).
    pub fn new(paramset: &str) -> Result<Self> {
        Ok(Self {
            inner: ffi::bin_new_context(paramset).map_err(err)?,
        })
    }

    /// `(n, q)` of the LWE ciphertexts this context accepts.
    pub fn lwe(&self) -> Result<(u64, u64)> {
        Ok((
            ffi::bin_lwe_n(&self.inner).map_err(err)?,
            ffi::bin_lwe_q(&self.inner).map_err(err)?,
        ))
    }

    pub fn load_keys(&mut self, refresh: &[u8], switching: &[u8]) -> Result<()> {
        ffi::bin_load_keys(self.inner.pin_mut(), refresh, switching).map_err(err)
    }

    pub fn load(&self, bytes: &[u8]) -> Result<BinCiphertext> {
        Ok(BinCiphertext {
            inner: ffi::bin_load(&self.inner, bytes).map_err(err)?,
        })
    }

    pub fn gate(&self, g: Gate, a: &BinCiphertext, b: &BinCiphertext) -> Result<BinCiphertext> {
        Ok(BinCiphertext {
            inner: ffi::bin_gate(&self.inner, g as u8, &a.inner, &b.inner).map_err(err)?,
        })
    }

    pub fn not(&self, a: &BinCiphertext) -> Result<BinCiphertext> {
        Ok(BinCiphertext {
            inner: ffi::bin_not(&self.inner, &a.inner).map_err(err)?,
        })
    }

    pub fn constant(&self, value: bool) -> Result<BinCiphertext> {
        Ok(BinCiphertext {
            inner: ffi::bin_constant(&self.inner, value).map_err(err)?,
        })
    }
}

impl BinCiphertext {
    pub fn store(&self) -> Result<Vec<u8>> {
        ffi::bin_store(&self.inner).map_err(err)
    }

    pub fn try_clone(&self) -> Result<Self> {
        Ok(Self {
            inner: ffi::bin_clone(&self.inner).map_err(err)?,
        })
    }
}

impl Clone for BinCiphertext {
    fn clone(&self) -> Self {
        self.try_clone().expect("OpenFHE BinFHE clone")
    }
}
